### PR TITLE
feat: Refactored Date and Date Range picker

### PR DESCRIPTION
**Pull request by albinAppsmith**
## Description

This PR includes refactored date and date range picker.

Fixes # (issue)
https://github.com/appsmithorg/appsmith/issues/19323
https://github.com/appsmithorg/appsmith/issues/19322

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual on storybook 
- Manual on main repo
- Jest
- Cypress

### Issues raised during DP testing
https://github.com/appsmithorg/design-system/pull/414#issuecomment-1469996724
https://github.com/appsmithorg/design-system/pull/414#issuecomment-1471903948
https://github.com/appsmithorg/design-system/pull/414#issuecomment-1473794440

### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: packages/design-system/src/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        [DateTimePickerClassName]: rest.showTimeInput,
+      })}
+      className={clsx(className, DatePickerClassName)}
+      customInput={
+        <Input
+          {...inputProps}
+          endIcon={
+            isClearable && selectedDate ? "close-circle-line" : undefined
+          }
+          endIconProps={{
```
Comment (tanvibhakta) [inline packages/design-system/src/DatePicker/DatePicker.tsx:69]: Didn't we decide we would never pass an onClick to an icon? If it needs an onClick then it should be an iconButton no?

File: packages/design-system/src/DatePicker/styles.css [hunk 2]
```diff
@@ -0,0 +1,165 @@
+.react-datepicker.ads-v2-datepicker__calender {
```
Comment (tanvibhakta) [inline packages/design-system/src/DatePicker/styles.css:1]: :( 

Can you leave a link to where the documentation for these classes are at the top of this file?

Comment (tanvibhakta) [reply packages/design-system/src/DatePicker/styles.css:1]: Also comments above each bit will be very nice.

**Review by tanvibhakta — CHANGES_REQUESTED** [review]

File: packages/design-system/src/DatePicker/DatePicker.tsx (hunk 1 above)
Comment (albinAppsmith) [reply packages/design-system/src/DatePicker/DatePicker.tsx:69]: we cannot use icon button here as the same as I mentioned for serach input. The variations and hover states and all needs to be sorted.

**Comment by shadabbuchh** [comment]
### Bugs in this PR:

**Functional:**

1. Deleting the JSON string in the excludesDates control does not reflect on the canvas when done in the `raw` mode. 
2. You can only enter 1 character at a time in the `yearStartRange` field. The cursor shifts after entering only one digit. Same happens with the `Label` field. (this happens intermittently)
3. The year dropdown always starts from the `yearStartRange`. This can be frustrating for the user who wants to select a relatively closer date. (suggestion is to show the selected year in the middle or something once the dropdown is clicked.)
4. Once we set the `yearEndRange` control, the year dropdown does not appear on clicking. Can we modify the desc so that it is not confusing for devs?
5. What is the difference between `isDisabled` & `isReadOnly`? I can select the date from the input in `isDisabled = true` state as well.
6. `allowSameDay` , `startDate` & `endDate` should not be there for Date picker.
7. Can the description of the `closeOnScroll` be a little more explanatory? (as in... scroll where?)
8. Default for `dateFormat` should be `mm/dd/yyyy` instead of `MM/dd/yyyy`. Also if we mention it as `mm/dd/yyyy` the month does not show up in the input field.
9. Not able to enter the date by typing in the input field. Also not able to delete the selected date without the `isClearable` control being set to `true`.
10. `selectedDate` does not work.
11. `timeFormat` does not work.
12. All time related fields/controls should not be there for Date Range Picker.
13. When `allowSameDay` is set to `false`, we are still able to select the same day in the Date Range Picker.
14. Even though it mentions that the `yearEndRange` default value is `current year + 1` but we can only see the current year as the End year in the dropdown.
15. The time value in `max/min Date` does not work & similarly the date fields in the `max/min Time` controls don't work. Can't we have controls that take both `max Date & Time` & `min Date & Time`? This is a better UX.
16. `max/min Time` don't work.
17. `Clear` & `Today` buttons are missing.

**Design:**

1. Wrong height & width of the whole card.
2. The height & width of the card changes on changing the month.
3. Wrong gap between month & year dropdown.
4. Gap between Days (Sun, Mon...) not visible. Should be 4px
5. Wrong gap between month/year dropdown & Days (Sun, Mon etc.)
6. Wrong colour for Dates.
7. The extreme left panel which displays `past week, month` etc. not visible.
8. The divider between months does not extend completely to the edges of the card.
9. Should the system date be shown in orange or what it is right now? (In design it is Orange)
10. Wrong font weight for month name & year.
11. Wrong box shadow for Month & year dropdown.

File: packages/design-system/src/DatePicker/DatePicker.tsx (hunk 1 above)
Comment (tanvibhakta) [reply packages/design-system/src/DatePicker/DatePicker.tsx:70]: Let's add a TODO here. https://theappsmith.slack.com/archives/D03K37HRU00/p1678879256621369?thread_ts=1678872894.044499&cid=D03K37HRU00

**Review by tanvibhakta — APPROVED** [review]

**Comment by albinAppsmith** [comment]
**Functional:**

1. Deleting the JSON string in the excludesDates control does not reflect on the canvas when done in the `raw` mode.  -> **Issue with storybook**
2. You can only enter 1 character at a time in the yearStartRange field. The cursor shifts after entering only one digit. Same happens with the Label field. (this happens intermittently)  -> **Issue with storybook**
3. The year dropdown always starts from the yearStartRange. This can be frustrating for the user who wants to select a relatively closer date. (suggestion is to show the selected year in the middle or something once the dropdown is clicked.) -> ** Will fix this later. Causing unexpected scroll behaviour. Create a separate bug for this.**
4. Once we set the yearEndRange control, the year dropdown does not appear on clicking. Can we modify the desc so that it is not confusing for devs? -> **Fixed**
5. What is the difference between isDisabled & isReadOnly? I can select the date from the input in isDisabled = true state as well. **As of now only focus state**
6. allowSameDay , startDate & endDate should not be there for Date picker. -> **Fixed**
7. Can the description of the closeOnScroll be a little more explanatory? (as in... scroll where?) -> **Fixed**
8. Default for dateFormat should be mm/dd/yyyy instead of MM/dd/yyyy. Also if we mention it as mm/dd/yyyy the month does not show up in the input field. **Fixed**
9. Not able to enter the date by typing in the input field. Also not able to delete the selected date without the isClearable control being set to true -> **This is expected**
10. selectedDate does not work. -> **Working for me, refresh the page once selected date is added https://design-system-git-ads-v2-date-picker-get-appsmith.vercel.app/?path=/story/design-system-datepicker--date-picker-story&args=selected:1678559400000**
11. timeFormat does not work. **Library is behaving differently, can you try date format.**
12. All time related fields/controls should not be there for Date Range Picker. -> **Done**
13. When allowSameDay is set to false, we are still able to select the same day in the Date Range Picker. -> **Library is not working according to this flag. Can you raise this as a separate issue and assign to me.**
14. Even though it mentions that the yearEndRange default value is current year + 1 but we can only see the current year as the End year in the dropdown. -> **Fixed**
15. The time value in max/min Date does not work & similarly the date fields in the max/min Time controls don't work. Can't we have controls that take both max Date & Time & min Date & Time? This is a better UX. -> **min max date is working for me. Regarding UI, that's not something we can control. It's part of storybook.https://design-system-git-ads-v2-date-picker-get-appsmith.vercel.app/?path=/story/design-system-datepicker--date-picker-story&args=maxDate:1678818600000;minDate:1678213800000**
16. max/min Time don't work. -> **Can you create a ticket and assign it to me**
17. Clear & Today buttons are missing. -> **Clear have a different UI in library, Can you create a separate ticket and assign it to me. **
18. For Date Picker, we can set maxDate > minDate. Can we prevent the users to do so? Same goes for maxTime & minTime -> **This is again something we cannot control in storybook.**
19. startDate & endDate do not work for Date Range Picker. -> **Working for me. Just refresh and try after updating. https://design-system-git-ads-v2-date-picker-get-appsmith.vercel.app/?path=/story/design-system-datepicker--date-range-picker-story&args=startDate:1677609000000;endDate:1678213800000**
20. Can we swap the placement of maxDate with minDate & similarly for maxTime & minTime. This is more intuitive. -> **Done**


1. Wrong height & width of the whole card. -> I **see it is mentioned as auto in design. Also this is taken from content. Can ignore this.**
2. The height & width of the card changes on changing the month. -> **Will fine tune later**
3,4,5 mention in new bug
6. Wrong colour for Dates. -> **Fixed**
7. The extreme left panel which displays past week, month etc. not visible. -> **This was a feature of blueprint, and the new library won't support this.**
8. The divider between months does not extend completely to the edges of the card. -> **We can ignore this for now. This was done to make sure it is consistant with other lines**
9. Should the system date be shown in orange or what it is right now? (In design it is Orange) -> **Fixed**
10. Wrong font weight for month name & year. -> **We are using the button tertiary component here, so it will have button tertiary font-weight.**
11. Wrong box shadow for Month & year dropdown. -> **Fixed**
12. We get to see a hover state for disabled dates (dates that cannot be selected due to the user selections). -> **Fixed**

**Comment by shadabbuchh** [comment]
Issues after the recent fix:

1. Wrong shade of orange applied to the system date.
2. Although the default dateFormat mentioned is `MM/dd/yyy` however I see that in the field `dd-MM-yyy` is selected. Can we make it uniform?

Other issues:

1. Can the dates for the month(s) that is not selected be in a lighter shade than #4C5664?

**Comment by shadabbuchh** [comment]
All identified issues are fixed. There is just one thing @albinAppsmith... In the Date Range Picker, when we move our cursor on days that do not have a date mentioned we get to see the hover state & we are also able to select the dates corresponding to those days.
https://www.loom.com/share/465a2563e66449cfb2762a0f9cd8a7da

**Comment by albinAppsmith** [comment]
> All identified issues are fixed. There is just one thing @albinAppsmith which doesn't seem to be fixed... In the Date Range Picker, when we move our cursor on days that do not have a date mentioned we get to see the hover state & we are also able to select the dates corresponding to those days. https://www.loom.com/share/465a2563e66449cfb2762a0f9cd8a7da

Will take this up in phase 2, since there is click possible in that empty columns.

**Pull request merged**